hazelcast/src/main/java/com/hazelcast/internal/util/IterableUtil.java

```diff
@@ -61,6 +69,72 @@ public void remove() {
         };
     }
 
+    /**
+     * Lazily filters iterated elements.
+     *
+     * @return a new iterable object which
+     * has an iterator capable of filtering elements
+     */
+    public static <T> Iterable<T> filter(Iterable<T> iterable, Predicate<T> filter) {
+        Iterator<T> givenIterator = iterable.iterator();
+
+        @SuppressWarnings("checkstyle:anoninnerlength")
+        Iterator<T> filteringIterator = new Iterator<T>() {
+            private T next;
+            private int hasNextCallCount;
```

Comment:
Not sure that we need these 2 variables. Why not using just `next`? If it is not `null` the hasNext() returns `true` immediately. The `next()` will first calls `hasNext()` and if it exists, returns it (assigning `next` to `null`). Something similar to the `HDBplusTree#EntryIterator`.

Reply:
but in that case consecutive `hasNext` calls advance state, i don't want it to happen.

Reply:
No, it'll check that `next` variable is not null and return quickly.

Reply:
Oops i see now, my test was wrong, also fixed it, thx.